Use cases voor oplossen verschillen (#110)
kiesraad/abacus-documentatie: use-cases/afhandeling-fouten-verschillen-coordinator.md

# Afhandeling fouten en verschillen door coördinator

Deze flowchart beschrijft hoe de coördinator (GSB of CSB) invoer met fouten of verschillen afhandelt.

Als de eerste invoer fouten bevat, moeten die opgelost worden. Dit betekent dat de tweede invoer alleen kan starten, nadat alle fouten die gedetecteerd zijn tijdens de eerste invoer, zijn opgelost.

Als de tweede invoer fouten bevat, heeft die dus per definitie verschillen met de eerste invoer. In dat geval moet de coördinator eerst de verschillen oplossen. Als de coördinator besluit de tweede invoer te behouden, dan moet als volgende stap de fouten opgelost worden.

> Bij het oplossen van verschillen kan de coördinator er, in plaats van de afwijkende invoer helemaal opnieuw te laten doen, ook voor kiezen om alleen de afwijkende velden te laten corrigeren door de oorspronkelijke invoerder van die invoer. Bij de correctie worden de velden met verschillen leeggemaakt en gemarkeerd met waarschuwing W.002. W.001 wordt bij een correctie-invoer niet getoond. Komt de gecorrigeerde invoer na de correctie nog steeds niet overeen met de bewaarde invoer, dan moet de coördinator de verschillen opnieuw oplossen.

Reviewer: Lijkt me goed om in use-cases weg te blijven van implementatie specifieke terminologie (zoals nummers van fouten en waarschuwingen)

```suggestion
Bij het oplossen van verschillen kan de coördinator er, in plaats van de afwijkende invoer helemaal opnieuw te laten doen, ook voor kiezen om alleen de afwijkende velden te laten corrigeren door de oorspronkelijke invoerder van die invoer. Bij de correctie worden de velden met verschillen leeggemaakt en gemarkeerd met waarschuwing. De invoerder kan in de correctie-invoer niet zien of nieuwe invoer afwijkt ten opzichte van de andere invoer. Komt de gecorrigeerde invoer na de correctie nog steeds niet overeen met de bewaarde invoer, dan moet de coördinator de verschillen opnieuw oplossen.
```

Correctie is alleen mogelijk voor een handmatige invoer. Een digitale invoer (bij het CSB is de eerste invoer geïmporteerd uit het tellingbestand van het GSB) kan niet gecorrigeerd worden, Abacus toont dan een melding. Alleen handmatig opnieuw invoeren is mogelijk.

> Correctie is verder alleen mogelijk als de invoer die bewaard wordt zelf geen fouten bevat. Bevat de te bewaren invoer fouten, dan biedt Abacus de correctie-optie niet aan en wordt er een melding getoond. Na bevestigen wordt de bewaarde invoer de eerste invoer en moeten de fouten daarin opgelost worden.

Reviewer: - Ook hier: use case zou ook moeten werken voor een andere implementatie dan Abacus. 
- Ik twijfel of het benoemen van 'bewaarde invoer wordt eerste invoer' in de use case thuishoort, aangezien dit een implementatiekeuze is.

```suggestion
Correctie is verder alleen mogelijk als de invoer die bewaard wordt zelf geen fouten bevat. Bevat de te bewaren invoer fouten, dan wordt de correctie-optie niet aangeboden en wordt er een melding getoond. Na bevestigen wordt de bewaarde invoer de eerste invoer en moeten de fouten daarin opgelost worden.
```

> Een gecorrigeerde invoer kan wél nieuwe fouten bevatten. Daardoor kan ook de eerste invoer fouten bevatten op het moment dat de coördinator verschillen oplost. Dit kon zonder de correctie-flow niet kon voorkomen, omdat fouten in de eerste invoer altijd direct na het afronden van de eerste invoer werden afgehandeld. Ook dan geldt de regel hierboven: de invoer met fouten kan niet bewaard worden in combinatie met een correctie van de tweede invoer.

Reviewer: ```suggestion
Een gecorrigeerde invoer kan wél nieuwe fouten bevatten. Daardoor kan ook de eerste invoer fouten bevatten op het moment dat de coördinator verschillen oplost. Ook dan geldt de regel hierboven: de invoer met fouten kan niet bewaard worden in combinatie met een correctie van de tweede invoer.
```

Is de implemtantatie historie relevant voor de use case?

    end

    invoer-met-fouten-en-waarschuwingen-2 --> verschil-oplossen
    verschil-oplossen -- bewaar 1ste invoer, laat 2de opnieuw invoeren --> klaar-voor-tweede-invoer

>     verschil-oplossen -- bewaarde 2de invoer, 2de invoer wordt 1ste invoer --> fouten

Reviewer 2: ```suggestion
    verschil-oplossen -- bewaar 2de invoer, 2de invoer wordt 1ste invoer --> fouten
```

>     verschil-oplossen -- bewaarde 2de invoer, 2de invoer wordt 1ste invoer --> fouten

Reviewer 2: Als je de 1ste invoer laat corrigeren, de gecorrigeerde eerste invoer fouten bevat, maar de coördinator beslist dat die invoer wel correct is, dan bewaar je de 1ste invoer in deze stap, niet de 2de invoer.

Author: Dan ga je toch via "bewaar 2de invoer, laat 1ste corrigeren" en daarna "fout oplossen"?

Reviewer 2: Ik heb de flow aangeduid op de flowchart:
<img width="909" height="1264" alt="image" src="https://github.com/user-attachments/assets/513e3943-fa25-4824-b67a-b8ba6ec5edef" />

Mijn punt is dat de tekst met de oranje pijl in dit geval niet correct is, want je bewaart daar de 1ste invoer.

Author: Ik denk dat ik je snap, heb het in a01d410f07db71727b6aa6a5dbe5716317795e9f proberen op te lossen

Reviewer 2: mooie oplossing!
